Exit Worker process on Fatal exceptions

JaredNeil opened the pull request:
The default global ExecutionContext simply logs uncaught exceptions. Since Future only catches NonFatal exceptions, any Fatal exceptions would cause the Future to never complete, which meant a WorkResponse was never sent back to the bazel server, which would then be stuck waiting forever for a response.

Since Fatal exceptions can leave the JVM in an invalid state, we exit if any thread has an unhandled Fatal exception.

Commit:
Exit Worker process on Fatal exceptions

The default global ExecutionContext simply logs uncaught exceptions.
Since Future only catches NonFatal exceptions, any Fatal exceptions
would cause the Future to never complete, which meant a WorkResponse was
never sent back to the bazel server, which would then be stuck waiting
forever for a response.

Since Fatal exceptions can leave the JVM in an invalid state, we exit if
any thread has an unhandled Fatal exception.

jroylance reviewed (approved):
LGTM, assuming it works when tested

Commit:
Catch and downgrade StackOverflowError

Zinc seems to run into this error pretty frequently, and if uncaught,
the entire worker process will exit suddenly, which can leave Bazel in a
weird state where it tries to send messages to the worker, but can't
because the process has already exited. Consider this a temporary
workaround until the underlying issues with Bazel can be fixed.
